e2e/config/vhd.go

```diff
@@ -0,0 +1,125 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute"
+)
+
+const imageGallery = "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.Compute/galleries/PackerSigGalleryEastUS/images/"
+
+var (
+	VHDUbuntu1804Gen2Containerd               = newVHDResourceIDFetcher(imageGallery + "1804Gen2")
+	VHDUbuntu2204Gen2Arm64Containerd          = newVHDResourceIDFetcher(imageGallery + "2204Gen2Arm64")
+	VHDUbuntu2204Gen2Containerd               = newVHDResourceIDFetcher(imageGallery + "2204Gen2")
+	VHDAzureLinuxV2Gen2Arm64                  = newVHDResourceIDFetcher(imageGallery + "AzureLinuxV2Gen2Arm64")
+	VHDAzureLinuxV2Gen2                       = newVHDResourceIDFetcher(imageGallery + "AzureLinuxV2Gen2")
+	VHDCBLMarinerV2Gen2Arm64                  = newVHDResourceIDFetcher(imageGallery + "CBLMarinerV2Gen2Arm64")
+	VHDCBLMarinerV2Gen2                       = newVHDResourceIDFetcher(imageGallery + "CBLMarinerV2Gen2")
+	VHDUbuntu2204Gen2ContainerdPrivateKubePkg = func() (VHDResourceID, error) {
+		return imageGallery + "2204Gen2PrivateKubePkg", nil
+	}
+)
+
+// VHDResourceID represents a resource ID pointing to a VHD in Azure. This could be theoretically
+// be the resource ID of a managed image or SIG image version, though for now this will always be a SIG image version.
+type VHDResourceID string
+
+// newVHDResourceIDFetcher is a factory function
+// it returns a function that fetches the latest VHDResourceID for a given image
+// the function is memoized and will only evaluate once on the first call
+func newVHDResourceIDFetcher(image string) func() (VHDResourceID, error) {
+	resourceID := VHDResourceID("")
+	var err error
+	once := sync.Once{}
+	// evaluate the function once and cache the result
+	return func() (VHDResourceID, error) {
+		once.Do(func() {
+			resourceID, err = findLatestImageWithTag(image, SIGVersionTagName, SIGVersionTagValue)
+		})
+		return resourceID, err
+	}
+}
+
+func (id VHDResourceID) Short() string {
+	sep := "Microsoft.Compute/galleries/"
+	str := string(id)
+	if strings.Contains(str, sep) && !strings.HasSuffix(str, sep) {
+		return strings.Split(str, sep)[1]
+	}
+	return str
+}
+
+var ErrNotFound = fmt.Errorf("not found")
+
+func findLatestImageWithTag(imageID, tagName, tagValue string) (VHDResourceID, error) {
```

Comment:
what would happen if we haven't run the VHD build pipeline in MSFT for at least 3 days and there are no longer any images that have the master branch tag due to the GC?

I guess we'd need to always keep at least one set of images locked like we currently do today so we can guarantee that the test suite is always runnable by itself

Reply:
one thing to note is that with the new duplication of AME images into MSFT galleries for RP e2e's, things will get a lot easier

once that effort is done (which it basically it is at this point, the galleries exist with actual images already), we can simplify all this logic to something like:

```
if tag name and value are specified:
  if tag is `buildId`
    take images from `PackerSigGalleryEastUS` that match the tag and value
 otherwise:
    take the latest images from `PackerSigGalleryEastUS` that match the tag and value
otherwise:
  take latest images from the new MSFT galleries containing copies of official AME images
```

the new galleries are named such as `AKSUbuntu`, `AKSAzureLinux`, `AKSCBLMariner`, and live within a separate DevInfra subscription within the MSFT tenant, which we'll most likely need to have access to via service principal

Reply:
I think something like this reflect better my intent:

```
if tag name and value are specified:
    take the latest images from `PackerSigGalleryEastUS` that match the tag and value. Skip tests if it doesn't exist
otherwise:
  take latest images from the new MSFT galleries containing copies of official AME images
```

Reply:
> what would happen if we haven't run the VHD build pipeline in MSFT for at least 3 days and there are no longer any images that have the master branch tag due to the GC?

How does it work currently? Wouldn't we have the same issue if GC deletes versions referenced in the e2e tests? What's the difference?

Reply:
today as a part of our "update agentbaker e2e VHD" process, we add a delete lock to the VHD's themselves to ensure they're protected from GC